src/TorchSharp/Distributions/Constraints.cs

```diff
@@ -0,0 +1,637 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace TorchSharp
+{
+
+    public static partial class torch
+    {
+        public static partial class distributions
+        {
+            public static partial class constraints
+            {
+                /// <summary>
+                /// Abstract base class for constraints.
+                /// 
+                /// A constraint object represents a region over which a variable is valid, e.g. within which a variable can be optimized.
+                /// </summary>
+                /// <remarks>
+                /// // It's not ideal to use a '_' first in a public .NET type name, but that's what Pytorch does for all contraint types.
+                /// </remarks>
+                public abstract class Constraint
+                {
+                    /// <summary>
+                    /// Constructor
+                    /// </summary>
+                    /// <param name="is_discrete">Whether constrained space is discrete.</param>
+                    /// <param name="event_dim">
+                    /// Number of rightmost dimensions that together define an event.
+                    /// The check() method will remove this many dimensions when computing validity.
+                    /// </param>
+                    protected Constraint(bool is_discrete = false, int event_dim = 0)
```

Comment:
I suggest adding default constructor something like:

protected Constraint() : this(false, 0) {}

then make this constructor with no default parameters 

protected Constraint(bool is_discrete, int event_dim) {...}

This to be explicit and avoid having the compiler adding the default constructor.

Reply:
Okay. I don't understand why this is useful/important, but I'll take your word for it.